trace/config.go

```diff
@@ -22,110 +22,160 @@ import (
-	Attributes []attribute.KeyValue
-	// Timestamp is a time in a Span life-cycle.
-	Timestamp time.Time
-	// Links are the associations a Span has with other Spans.
-	Links []Link
-	// NewRoot identifies a Span as the root Span for a new trace. This is
-	// commonly used when an existing trace crosses trust boundaries and the
-	// remote parent span context should be ignored for security.
-	NewRoot bool
-	// SpanKind is the role a Span has in a trace.
-	SpanKind SpanKind
-}
-
-// NewSpanConfig applies all the options to a returned SpanConfig.
+	attributes []attribute.KeyValue
+	timestamp  time.Time
+	links      []Link
+	newRoot    bool
+	spanKind   SpanKind
+}
+
+// Attributes describe the associated qualities of a Span.
+func (cfg *SpanConfig) Attributes() []attribute.KeyValue {
+	return cfg.attributes
+}
+
+// Timestamp is a time in a Span life-cycle.
+func (cfg *SpanConfig) Timestamp() time.Time {
+	return cfg.timestamp
+}
+
+// Links are the associations a Span has with other Spans.
+func (cfg *SpanConfig) Links() []Link {
+	return cfg.links
+}
+
+// NewRoot identifies a Span as the root Span for a new trace. This is
+// commonly used when an existing trace crosses trust boundaries and the
+// remote parent span context should be ignored for security.
+func (cfg *SpanConfig) NewRoot() bool {
+	return cfg.newRoot
+}
+
+// SpanKind is the role a Span has in a trace.
+func (cfg *SpanConfig) SpanKind() SpanKind {
+	return cfg.spanKind
+}
+
+// NewSpanStartConfig applies all the options to a returned SpanConfig.
 // No validation is performed on the returned SpanConfig (e.g. no uniqueness
 // checking or bounding of data), it is left to the SDK to perform this
 // action.
-func NewSpanConfig(options ...SpanOption) *SpanConfig {
+func NewSpanStartConfig(options ...SpanOption) *SpanConfig {
 	c := new(SpanConfig)
 	for _, option := range options {
-		option.ApplySpan(c)
+		option.applySpanStart(c)
 	}
 	return c
 }
 
-// SpanOption applies an option to a SpanConfig.
+// NewSpanEndConfig applies all the options to a returned SpanConfig.
+// No validation is performed on the returned SpanConfig (e.g. no uniqueness
+// checking or bounding of data), it is left to the SDK to perform this
+// action.
+func NewSpanEndConfig(options ...SpanEndOption) *SpanConfig {
+	c := new(SpanConfig)
+	for _, option := range options {
+		option.applySpanEnd(c)
+	}
+	return c
+}
+
+// SpanOptions applies an option to a SpanConfig. These options are applyable
+// Only when the span is created
 type SpanOption interface {
-	ApplySpan(*SpanConfig)
+	applySpanStart(*SpanConfig)
+}
+
+type SpanOptionFunc func(*SpanConfig)
 
-	// A private method to prevent users implementing the
-	// interface and so future additions to it will not
-	// violate compatibility.
-	private()
+func (fn SpanOptionFunc) applySpanStart(cfg *SpanConfig) {
+	fn(cfg)
+}
+
+type SpanEndOption interface {
+	applySpanEnd(*SpanConfig)
+}
+
+type EventConfig struct {
```

Comment:
Missing documentation.

```suggestion
// EventConfig is a group of options for an Event.
type EventConfig struct {
```